add rottweiler to manage storage encryption

bcressey opened the pull request:
**Issue number:**
Related: https://github.com/bottlerocket-os/bottlerocket/issues/4680

**Description of changes:**
With Secure Boot enabled and dm-verity enforced, we usually consider the kernel and root filesystem to be trusted. Encrypting BOTTLEROCKET-DATA (`/local`) and BOTTLEROCKET-PRIVATE (`/.bottlerocket`) helps expand that trust boundary to include mutable storage as well, which is useful for remote attestation.

We can encrypt the entire data partition as we usually expect it to be empty and create the filesystem on first boot. However, the private partition cannot be fully encrypted; GRUB needs to read the bootconfig file, and there's potentially also a `net.toml` for network configuration. That means two types of encryption to support - LUKS for full disk encryption, fscrypt for directories.

`systemd` has excellent support for encrypted block devices, but not for directories except when `systemd-homed` is in use. [Managing encrypted filesystems with dirlock](https://lwn.net/Articles/1038859/) provides an overview of some tools in the space, with `dirlock` as an interesting project. However, most of the tools are opinionated (and flexible!) in ways that don't mesh perfectly with our needs. In particular, any sort of automatic fallback to passwords is problematic; if the storage cannot be automatically unlocked, then it can't be trusted and the system should fail closed.

`rottweiler` exists to provide a uniform wrapper around the underlying tools: `systemd-creds`, `cryptsetup`, and the kernel's fscrypt API. It requires a TPM2 device, and uses `systemd-creds` to generate wrapped keys that are sealed to specific PCRs. 


**Testing done:**
Enabled the [encrypted storage](https://github.com/bottlerocket-os/twoliter/pull/589) feature. Booted (and rebooted) images with encrypted block devices and directories.

That exercises these code paths:
```
# /local
rottweiler check block-device ${BOTTLEROCKET_DATA} unencrypted
rottweiler generate-key bottlerocket-data
rottweiler encrypt block-device ${BOTTLEROCKET_DATA} bottlerocket-data
rottweiler attach block-device ${BOTTLEROCKET_DATA} bottlerocket-data
rottweiler detach block-device ${BOTTLEROCKET_DATA}

# /.bottlerocket/datastore
rottweiler check directory ${DATASTORE_DIR} unencrypted
rottweiler generate-key datastore
rottweiler encrypt directory ${DATASTORE_DIR} datastore
```

I've also tested the lock and unlock paths:
```
$ rottweiler lock directory /.bottlerocket/datastore
<succeeds>

# file names are encrypted
$ ls -1 /.bottlerocket/datastore/
8Dpj4eE5GltMa8f00gibrGAoPQGRwM6dpl1KVTJwAHDUs_XVrilBeA
DiPPFeTtYxFN6kBKUkbnRijqId6IlecykRDKSuaFXoT7VOvnX2nswA
UJS1XpONYwoewy4cD9S68nrKCrozqkr33B_G8RnF-Ma6c-GeAbRKKA
eouecYx9ArlXpKsgZrYaDy7bxWrM_AlcXDVmsP0JEGgjjBr-pfgXJw
jM5WLUKW_72Ir3AAqm_Fx9dlekNo4NdO8PVlKhPbukAuu-suNmbaSw
nqDxnwOaczPa3kkYp_14aawGEpszfOYNuidNNRMBNKv25AiuvEcVAA

# files are inaccessible
$ cat /.bottlerocket/datastore/DiPPFeTtYxFN6kBKUkbnRijqId6IlecykRDKSuaFXoT7VOvnX2nswA
cat: /.bottlerocket/datastore/DiPPFeTtYxFN6kBKUkbnRijqId6IlecykRDKSuaFXoT7VOvnX2nswA: Required key not available

# directory can't be unlocked if the PCRs aren't in the expected state
$ rottweiler unlock directory /.bottlerocket/datastore datastore
Error: /usr/bin/systemd-creds failed with exit code 1: Failed to unseal secret using TPM2: Operation not permitted
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

Commit:
selinux-policy: let rottweiler write to private fs

`rottweiler` needs to write to BOTTLEROCKET-PRIVATE in order to make
the datastore directory and set its encryption policy.

It also stores its keys on that filesystem.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
sources: add rottweiler

rottweiler provides a uniform wrapper around the standard underlying
tools: systemd-creds, cryptsetup, and the kernel's fscrypt API.

It requires a TPM2 device, and uses systemd-creds to generate wrapped
keys that are sealed to specific PCRs.

It doesn't support any form of interactive authentication or fallback
to keys stored outside the TPM2.

It also provides "measure" commands to ensure consistent measurements
of the kernel command line, OS settings, and boot phases.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

arnaldo2792 commented on packages/rottweiler/rottweiler.spec:
```diff
@@ -0,0 +1,31 @@
+
+%build
+%cargo_build --manifest-path %{_builddir}/sources/Cargo.toml \
+    -p rottweiler
```
I assume the `cross_scan_attribution` isn't needed because of this?

https://github.com/bottlerocket-os/bottlerocket-core-kit/blob/cfe9fded8063b79402e517077b11f77596bdcd73/packages/os/os.spec#L750

bcressey replied:
> I assume the `%cross_scan_attribution` isn't needed because of this?

Yes, we'll create the right attribution files based on the scan in `os`.

cbgbt commented on sources/rottweiler/src/key.rs:
```diff
@@ -0,0 +1,51 @@
+        .read_exact(&mut random_bytes)
+        .with_whatever_context(|_| "failed to read random bytes")?;
+
+    let encrypted = system::systemd_creds_encrypt(&key_id, &random_bytes)?;
+
```
We should potentially consider [zeroize](https://docs.rs/zeroize/latest/zeroize/)ing the random bytes here to force them out of process memory once the encryption is complete.

cbgbt commented on sources/rottweiler/src/key.rs:
```diff
@@ -0,0 +1,51 @@
+    let encrypted = fs::read(&key_path)
+        .with_whatever_context(|_| format!("failed to read key from '{}'", key_path.display()))?;
+
+    system::systemd_creds_decrypt(&key_id, &encrypted)
```
Could return a `Zeroizing<T>` wrapper here.

cbgbt commented on sources/rottweiler/src/system.rs:
```diff
@@ -0,0 +1,207 @@
+}
+
+/// Decrypt data using systemd-creds with TPM2
+pub fn systemd_creds_decrypt(name: &str, ciphertext: &[u8]) -> Result<Vec<u8>> {
```
Likewise, could opt to return a `Zeroizing<T>` type here to ensure the returned creds are removed from memory as soon as possible.

cbgbt commented on sources/rottweiler/src/system.rs:
```diff
@@ -0,0 +1,207 @@
+}
+
+/// Execute a command with optional stdin input and return stdout
+fn execute(cmd: &str, args: &[&str], input: Option<&[u8]>) -> Result<Vec<u8>> {
```
We should consider zero-izing the stdin and stdout buffers.

cbgbt commented on sources/rottweiler/src/fscrypt.rs:
```diff
@@ -0,0 +1,291 @@
+        Ok(())
+    }
+}
+
+/// Private key for fscrypt operations (contains the key identifier and raw key material)
+pub struct FscryptPrivateKey {
```
Could derive the `zeroize` type here.

cbgbt commented on sources/rottweiler/src/key.rs:
```diff
@@ -0,0 +1,51 @@
+    if key_path.exists() {
+        return Ok(());
+    }
+
```
Do we care about potential TOCTOU issues here?

cbgbt commented on sources/rottweiler/src/mount_point.rs:
```diff
@@ -0,0 +1,64 @@
+/// stat() returns dev_t as u64 using old encoding: (major & 0xff) << 8 | (minor & 0xff) | ((minor & 0xfff00) << 12)
+/// Kernel uses new encoding internally: (major << 20) | minor, where major fits in 12 bits and minor in 20 bits
+/// This function extracts major/minor from old encoding and re-encodes to match kernel's format.
+fn dev_to_kernel_encoding(stat_dev: u64) -> u32 {
```
Pure function probably warrants a few tests.

cbgbt commented on sources/rottweiler/src/key.rs:
```diff
@@ -0,0 +1,51 @@
+
+/// Generate a random encryption key and encrypt it with TPM2 PCRs 7+14
+pub fn generate(key_id: String) -> Result<()> {
+    let key_path = PathBuf::from(KEYSTORE).join(&key_id);
```
We could consider sanitizing `key_id` for path traversal elements, but as I understand it, the input here is almost certainly coming from a static systemd service, so I don't think that there's necessarily a problem here.

cbgbt reviewed (commented):
Still working on a more thorough review, but here are some initial ideas from stepping through the code with Q.

cbgbt commented on sources/rottweiler/src/block_device.rs:
```diff
@@ -0,0 +1,93 @@
+        .with_whatever_context(|| {
+            format!("volume name is not valid UTF-8: '{}'", path.display())
+        })?;
+
```
Consider extracting to a function, given the repetition.

cbgbt commented on sources/rottweiler/src/system.rs:
```diff
@@ -0,0 +1,207 @@
+    execute(
+        CRYPTSETUP,
+        &[
+            "luksFormat",
```
I'd wondered if detecting the physical sector size of the underlying device is a good idea. Instance types with 4K physical sectors could perform better if we format to the same sector size.

On the contrary, though, if you re-use an EBS snapshot you could end up in a situation where the new host has a smaller physical sector size. My understanding is that having larger LUKS sectors than the physical disk sectors could lead to data corruption on power loss.

bcressey replied:
> On the contrary, though, if you re-use an EBS snapshot you could end up in a situation where the new host has a smaller physical sector size. My understanding is that having larger LUKS sectors than the physical disk sectors could lead to data corruption on power loss.

Since the encryption key is sealed to the host's TPM, it won't be available on a different host, meaning it won't be possible to mount the LUKS volume elsewhere.

cbgbt commented on sources/rottweiler/src/system.rs:
```diff
@@ -0,0 +1,207 @@
+    let content = fs::read_to_string(IMAGE_FEATURES_FILE)
+        .with_whatever_context(|_| format!("failed to read {}", IMAGE_FEATURES_FILE))?;
+
+    let pairs: Vec<(String, String)> = content
```
We should probably add unit tests for this parser.

cbgbt commented on sources/rottweiler/src/mount_point.rs:
```diff
@@ -0,0 +1,64 @@
+        let current_stat = std::fs::metadata(&current).with_whatever_context(|_| {
+            format!("failed to get metadata for '{}'", current.display())
+        })?;
+        let current_dev = dev_to_kernel_encoding(current_stat.dev());
```
Do we actually want to do this encoding transformation? We're always comparing devices as returned by rust's stdlib, so the kernel encoding should not matter.

I'm pretty sure we can remove the `dev_to_kernel_encoding` function.

bcressey replied:
> I'm pretty sure we can remove the dev_to_kernel_encoding function.

You're right. This is an artifact of the BPF LSM hook that I'm planning to add later, to block changes to a mount point.

cbgbt commented on sources/rottweiler/src/mount_point.rs:
```diff
@@ -0,0 +1,64 @@
+/// This function extracts major/minor from old encoding and re-encodes to match kernel's format.
+fn dev_to_kernel_encoding(stat_dev: u64) -> u32 {
+    // Extract major and minor from old stat encoding
+    let major = ((stat_dev >> 8) & 0xff) as u32; // 8 bits max
```
I think the major number can actually be > 8 bits now. NVME devices have a major of 259.

But I left another comment pointing out that we probably don't actually need to change the device number encoding.

cbgbt reviewed (commented):
Still working through directory encryption!

cbgbt commented on sources/rottweiler/src/fscrypt.rs:
```diff
@@ -0,0 +1,291 @@
+        let dir_fd = std::fs::File::open(path)
+            .with_whatever_context(|_| format!("Failed to open directory '{}'", path.display()))?;
+        let argptr = &mut policy as *mut FscryptPolicyV2 as *mut FscryptPolicyV1Ioctl;
+        unsafe { set_encryption_policy(dir_fd.as_raw_fd(), argptr) }.with_whatever_context(
+            |_| format!("Failed to set encryption policy for '{}'", path.display()),
+        )?;
```
```suggestion
        // SAFETY: The ioctl requires a valid file descriptor and a pointer to a policy struct.
        // - `dir_fd` is a valid open file descriptor for the directory
        // - The kernel reads the version field first (via get_user), then copies the full struct
        // - Both FscryptPolicyV2 and FscryptPolicyV1Ioctl are repr(C) and start with version field
        // - The cast is safe because the kernel interprets the data based on the version field
 
        unsafe { set_encryption_policy(dir_fd.as_raw_fd(), argptr) }.with_whatever_context(
            |_| format!("Failed to set encryption policy for '{}'", path.display()),
        )?;
```

cbgbt commented on sources/rottweiler/src/fscrypt.rs:
```diff
@@ -0,0 +1,291 @@
+        })?;
+        let mount_fd = mount_point.open()?;
+        let mut remove_key = FscryptRemoveKey::new(self.identifier);
+        unsafe { remove_encryption_key_all_users(mount_fd.as_raw_fd(), &mut remove_key) }
```
```suggestion
        // SAFETY: The ioctl requires a valid file descriptor and a pointer to fscrypt_remove_key_arg.
        // - `mount_fd` is a valid open file descriptor
        // - `remove_key` is properly initialized with repr(C) layout matching the kernel struct
        // - The kernel copies data with copy_from_user/copy_to_user, so no lifetime issues
        unsafe { remove_encryption_key_all_users(mount_fd.as_raw_fd(), &mut remove_key) }
```

cbgbt commented on sources/rottweiler/src/fscrypt.rs:
```diff
@@ -0,0 +1,291 @@
+        let mount_fd = mount_point.open()?;
+        let mut add_key = FscryptAddKey::new(&self.raw)?;
+        let argptr = &mut add_key as *mut FscryptAddKey as *mut FscryptAddKeyIoctl;
+        unsafe { add_encryption_key(mount_fd.as_raw_fd(), argptr) }.with_whatever_context(
+            |_| format!("Failed to add encryption key for '{}'", path.display()),
+        )?;
```
```suggestion
        // SAFETY: The ioctl requires a valid file descriptor and a pointer to fscrypt_add_key_arg.
        // - `mount_fd` is a valid open file descriptor for the mount point
        // - `add_key` is properly initialized with repr(C) layout matching the kernel struct
        // - The cast is safe: FscryptAddKey contains FscryptAddKeyIoctl fields plus raw[] array
        // - The kernel uses copy_from_user to read the struct including the raw key bytes
        unsafe { add_encryption_key(mount_fd.as_raw_fd(), argptr) }.with_whatever_context(
            |_| format!("Failed to add encryption key for '{}'", path.display()),
        )?;
```

cbgbt commented on sources/rottweiler/src/fscrypt.rs:
```diff
@@ -0,0 +1,291 @@
+            .with_whatever_context(|_| format!("Failed to open directory '{}'", path.display()))?;
+        let mut arg = Self::new()?;
+        let argptr = &mut arg as *mut FscryptGetPolicyExArg as *mut FscryptGetPolicyExArgIoctl;
+        unsafe { get_encryption_policy_ex(dir.as_raw_fd(), argptr) }
+            .with_whatever_context(|_| "Failed to get encryption policy")?;
```
```suggestion
        // SAFETY: The ioctl requires a valid file descriptor and a pointer to fscrypt_get_policy_ex_arg.
        // - `dir` is a valid open file descriptor for the directory
        // - `arg` is properly initialized with policy_size and version fields
        // - The kernel first copies policy_size from userspace, then copies the full policy back
        // - Both structs are repr(C) and share the same initial layout (policy_size field)
        unsafe { get_encryption_policy_ex(dir.as_raw_fd(), argptr) }
            .with_whatever_context(|_| "Failed to get encryption policy")?;
```

cbgbt commented on sources/rottweiler/src/fscrypt.rs:
```diff
@@ -0,0 +1,291 @@
+    }
+
+    fn key_identifier(&self) -> [u8; 16] {
+        unsafe { self.policy.v2.master_key_identifier }
```
```suggestion
        // SAFETY: Accessing union field is safe because:
        // - We initialized the policy union with version = FSCRYPT_POLICY_V2
        // - The kernel's fscrypt_ioctl_get_policy_ex writes the policy based on what's stored
        // - The kernel implementation (fs/crypto/policy.c) guarantees the v2 variant is active
        //   when the stored policy version is v2
        // - We only call this after successfully retrieving the policy via ioctl
        unsafe { self.policy.v2.master_key_identifier }
```

But we could also modify the implementation to make it impossible to call this in Rust without first using a separate type. [Sample of the idea](https://gist.github.com/cbgbt/fb7366b36ceb5f554f972bef3be0eca4)

cbgbt commented on sources/rottweiler/src/fscrypt.rs:
```diff
@@ -0,0 +1,291 @@
+    version: u8,
+    contents_encryption_mode: u8,
+    filenames_encryption_mode: u8,
+    flags: u8,
+    __reserved: [u8; 4],
```
It seems like kernel 6.12 has eaten one byte of the reserved space here to add a field called `log2_data_unit_size`.

But`0` is still a reasonable default for that (use filesystem block size), so it doesn't have an impact on our implementation. I'm OK to leave the reserved space as-is.

cbgbt commented on sources/rottweiler/src/fscrypt.rs:
```diff
@@ -0,0 +1,291 @@
+    hkdf.expand(b"fscrypt\x00\x01", &mut output)
+        .with_whatever_context(|_| "Failed to derive key identifier")?;
+    Ok(output)
+}
```
I got Q to generate const assertions that the FFI types here match the layout of the kernel ones:


```rust
// Compile-time validation of struct layouts against kernel definitions
// These assertions ensure our repr(C) structs match the kernel's C structs exactly.

const _: () = {
    use std::mem::{align_of, offset_of, size_of};

    // struct fscrypt_policy_v1: 1 + 1 + 1 + 1 + 8 = 12 bytes
    const _: () = assert!(size_of::<FscryptPolicyV1Ioctl>() == 12);
    const _: () = assert!(align_of::<FscryptPolicyV1Ioctl>() == 1);

    // struct fscrypt_policy_v2: 1 + 1 + 1 + 1 + 1 + 3 + 16 = 24 bytes
    const _: () = assert!(size_of::<FscryptPolicyV2>() == 24);
    const _: () = assert!(align_of::<FscryptPolicyV2>() == 1);
    const _: () = assert!(offset_of!(FscryptPolicyV2, version) == 0);
    const _: () = assert!(offset_of!(FscryptPolicyV2, contents_encryption_mode) == 1);
    const _: () = assert!(offset_of!(FscryptPolicyV2, filenames_encryption_mode) == 2);
    const _: () = assert!(offset_of!(FscryptPolicyV2, flags) == 3);
    const _: () = assert!(offset_of!(FscryptPolicyV2, __reserved) == 4);
    const _: () = assert!(offset_of!(FscryptPolicyV2, master_key_identifier) == 8);

    // struct fscrypt_key_specifier: 4 + 4 + 32 = 40 bytes
    const _: () = assert!(size_of::<FscryptKeySpecifier>() == 40);
    const _: () = assert!(align_of::<FscryptKeySpecifier>() == 4);
    const _: () = assert!(offset_of!(FscryptKeySpecifier, type_) == 0);
    const _: () = assert!(offset_of!(FscryptKeySpecifier, __reserved) == 4);
    const _: () = assert!(offset_of!(FscryptKeySpecifier, u) == 8);

    // struct fscrypt_add_key_arg (without raw[]): 40 + 4 + 4 + 32 = 80 bytes
    const _: () = assert!(size_of::<FscryptAddKeyIoctl>() == 80);
    const _: () = assert!(align_of::<FscryptAddKeyIoctl>() == 4);
    const _: () = assert!(offset_of!(FscryptAddKey, key_spec) == 0);
    const _: () = assert!(offset_of!(FscryptAddKey, raw_size) == 40);
    const _: () = assert!(offset_of!(FscryptAddKey, key_id) == 44);
    const _: () = assert!(offset_of!(FscryptAddKey, __reserved) == 48);
    const _: () = assert!(offset_of!(FscryptAddKey, raw) == 80);

    // struct fscrypt_add_key_arg with raw[64]: 80 + 64 = 144 bytes
    const _: () = assert!(size_of::<FscryptAddKey>() == 144);
    const _: () = assert!(align_of::<FscryptAddKey>() == 4);

    // struct fscrypt_remove_key_arg: 40 + 4 + 20 = 64 bytes
    const _: () = assert!(size_of::<FscryptRemoveKey>() == 64);
    const _: () = assert!(align_of::<FscryptRemoveKey>() == 4);
    const _: () = assert!(offset_of!(FscryptRemoveKey, key_spec) == 0);
    const _: () = assert!(offset_of!(FscryptRemoveKey, removal_status_flags) == 40);
    const _: () = assert!(offset_of!(FscryptRemoveKey, __reserved) == 44);

    // struct fscrypt_get_policy_ex_arg: 8 + max(1, 12, 24) = 32 bytes
    const _: () = assert!(size_of::<FscryptGetPolicyExArg>() == 32);
    const _: () = assert!(align_of::<FscryptGetPolicyExArg>() == 8);

    // __u8[9] for ioctl definition: 8 + 1 = 9 bytes (packed)
    const _: () = assert!(size_of::<FscryptGetPolicyExArgIoctl>() == 9);
    const _: () = assert!(align_of::<FscryptGetPolicyExArgIoctl>() == 1);
};

```

KCSesh replied:
Good idea!

cbgbt reviewed (commented):
Done! Nice work.

Commit:
sources: add bottlerocket-image-features

Provide a standard way to parse the image-features.env file written
by Twoliter during image builds.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

cbgbt committed:
rottweiler: robustify fscrypt

Provide safety rationale for all unsafe blocks.

Add const expressions to ensure that struct and union definitions are
aligned with the kernel UAPI.

Signed-off-by: Ben Cressey <bcressey@amazon.com>

bcressey commented:
Added commits to fix all suggestions - thanks @cbgbt!

bcressey commented:
Added two new commits:
1. The new `measure` command works around a problem I was having with `tpm2_pcrevent`, and also cleans up a lot of clunky mechanics around setting up the input for that command in the absence of a shell.
2. The SELinux policy change came over here since it's small, and since `rottweiler` is managing its keystore directory on the private filesystem.

bcressey commented:
Added a commit to add parsing for the encrypted storage image feature too.

bcressey commented:
Added a fix to shift dependencies to `rottweiler.spec`, now that it wraps all of the underlying system commands.

cbgbt commented on sources/rottweiler/src/system.rs:
```diff
@@ -158,3 +160,33 @@ fn get_tpm2_pcrs() -> Result<String> {
     })
 }
+
+/// Get canonicalized settings from apiclient, excluding random-seed and hostname
```
It feels somewhat brittle to explicitly exclude these here, but I don't think we really have a feasible alternative.

I'd be OK with opening an issue to discuss, but perhaps some system where the variant owner can record specific settings in a file to mark them as being transient/host-identifying.

KCSesh replied:
Yeah I think I would want to see this built into apiclient as a subcommand? 
Maybe something like:

```
apiclient get-canonicalize settings
```
🤷 

It would be nice to have a test to help capture this, but I agree it makes sense for now giving no great alternatives.

bcressey replied:
Opened https://github.com/bottlerocket-os/bottlerocket-core-kit/issues/739 for this as a follow-up feature.

arnaldo2792 commented on sources/rottweiler/src/key.rs:
```diff
@@ -0,0 +1,51 @@
+pub fn generate(key_id: String) -> Result<()> {
+    let key_path = PathBuf::from(KEYSTORE).join(&key_id);
+
+    // Skip generation if key already exists
```
```suggestion

```

arnaldo2792 commented on sources/rottweiler/src/system.rs:
```diff
@@ -0,0 +1,207 @@
+///
+/// If in-place updates are disabled, then the measurements in PCR 4 (kernel) and
+/// PCR 9 (userspace) shouldn't change.
+fn get_tpm2_pcrs() -> Result<String> {
```
Were these PCRs arbitrarily selected?

bcressey replied:
> Were these PCRs arbitrarily selected?

No 😀 A lot of thought went into them. There's some documentation over in #721.

7 is Secure Boot; if the system isn't in Secure Boot mode, or if that ever changes for some reason, we don't want the sealed key to be available.

11 is extended whenever we finish a boot phase. Including it means that the sealed key can only ever be retrieved during the same boot phase that we first generated it.

14 is managed by shim and records machine-owner keys. There shouldn't ever be any, but if there are, it effectively means the set of keys trusted for Secure Boot has changed, which is unexpected and basically the same as Secure Boot being disabled under our threat model.

arnaldo2792 commented on sources/bottlerocket-image-features/src/lib.rs:
```diff
@@ -53,6 +53,8 @@ const IMAGE_FEATURES_FILE: &str = "/usr/share/bottlerocket/image-features.env";
 pub struct ImageFeatures {
```
Did you mean to `fixup` this commit?

bcressey replied:
> Did you mean to fixup this commit?

I'm planning to squash all commits together prior to merge. I wanted to make it easier for reviewers to see incremental changes in case they were partway through a review.

arnaldo2792 commented on packages/rottweiler/rottweiler.spec:
```diff
@@ -9,6 +9,8 @@ License: Apache-2.0 OR MIT
 URL: https://github.com/bottlerocket-os/bottlerocket
```
Same with this? `fixup!` the commit?

arnaldo2792 commented on sources/bottlerocket-image-features/README.md:
```diff
@@ -17,6 +17,7 @@ file, which contains feature flags that control various aspects of the Bottleroc
 Currently supported feature flags:
```
Same as elsewhere about the `fixup` commits

KCSesh commented on sources/rottweiler/src/system.rs:
```diff
@@ -0,0 +1,192 @@
+            "get",
+            "settings",
+            "--exclude",
+            "settings.updates.random-seed",
```
I am trying to think through the possibilities here across bootstrap commands/containers or host-containers setting values on boot. 

What was your methodology here for this list?

bcressey replied:
> What was your methodology here for this list?

I targeted settings that were different across two runs of the same AMI.

KCSesh commented on sources/rottweiler/src/system.rs:
```diff
@@ -0,0 +1,192 @@
+            "get",
+            "settings",
+            "--exclude",
+            "settings.updates.random-seed",
```
Am I missing an alias? 

```suggestion
            "settings.updates.seed",
```

KCSesh replied:
Seems this was just fixed 👍

bcressey replied:
> Am I missing an alias?

No, this was a bug. Too many random numbers on my mind 😀

bcressey commented:
Two fixes:
1. `settings.updates.seed` is the correct setting for update waves
2. when encrypting a directory, remove it first. The fscrypt API to set the encryption policy will fail if the directory is not empty, but this way the behavior is the same as when encrypting a block device (existing content is wiped).

bcressey commented:
⬆️ (hopefully) final force push to squash commits and finalize commit messages

KCSesh reviewed (approved):
Finalized changes LGTM
